### PR TITLE
Add Nginx README

### PR DESCRIPTION
**Pull request by alexandrevicenzi**
This is related to #993.

### CODE REVIEW

**Comment by alexandrevicenzi** [comment]
This is a Draft because I found #1123 and #547 to be a blocker for the proper use of this image that I described in the README.

I'll fix both issues, ideally, we should merge this after both are merged.

File: src/bci_build/package/nginx/README.md.j2
Comment (dcermak) [inline src/bci_build/package/nginx/README.md.j2:25]:
```suggestion
You can access the hosted pages by opening http://localhost:8080 or http://host-ip:8080 in your browser.
```

File: src/bci_build/package/nginx/README.md.j2
Comment (dcermak) [inline src/bci_build/package/nginx/README.md.j2:39]:
```suggestion
Out of the box, nginx doesn't support environment variables inside configuration blocks. This image ships a script that can extract environment variables before nginx starts.
```

File: src/bci_build/package/nginx/README.md.j2
Comment (dcermak) [inline src/bci_build/package/nginx/README.md.j2:41]:
```suggestion
The script reads template files in `/etc/nginx/templates/*.template` and outputs the result of executing `envsubst` to the directory `/etc/nginx/conf.d/`.
```

File: src/bci_build/package/nginx/README.md.j2
Comment (dcermak) [inline src/bci_build/package/nginx/README.md.j2:49]:
```suggestion
The script will render the template to `/etc/nginx/conf.d/default.conf` given the environment variable `NGINX_PORT` set to `80` as follows:
```

File: src/bci_build/package/nginx/README.md.j2
Comment (dcermak) [inline src/bci_build/package/nginx/README.md.j2:55]: This whole section should go above or below the environment variable section

File: src/bci_build/package/nginx/README.md.j2
Comment (dcermak) [inline src/bci_build/package/nginx/README.md.j2:67]: Does this actually work on SLES? I guess you'd have to adjust the load path for nginx config files as well, right?

File: src/bci_build/package/nginx/README.md.j2
Comment (dcermak) [inline src/bci_build/package/nginx/README.md.j2:85]: What are these stream templates? Something like the "ordinary" templates but with a different destination?

File: src/bci_build/package/nginx/README.md.j2
Comment (dcermak) [inline src/bci_build/package/nginx/README.md.j2:93]:
```suggestion
If you would like to use scripts to perform configurations, add one or more `*.envsh` or `*.sh` scripts under `/docker-entrypoint.d/`. Any executable `*.envsh` or `*.sh` script found in the directory will be executed before starting the service and can be used to perform further configuration steps.
```

File: src/bci_build/package/nginx/README.md.j2
Comment (dcermak) [inline src/bci_build/package/nginx/README.md.j2:95]:
```suggestion
Currently, the container image ships the following helper scripts:
```

File: src/bci_build/package/nginx/README.md.j2
Comment (dcermak) [inline src/bci_build/package/nginx/README.md.j2:101]:
```suggestion
**Warning:** The container startup is aborted if any of the script exits with an error.
```

**Review by dcermak — COMMENTED** [review]
Great README :100: 

I've added a few suggestions, but would leave the full language review to @dmpop, once you deem this ready.

File: src/bci_build/package/nginx/README.md.j2
Comment (alexandrevicenzi) [reply src/bci_build/package/nginx/README.md.j2:67]: Yes, probably needs a note here mentioning that if you change this path, you must change nginx.conf otherwise it will still load the old path.

File: src/bci_build/package/nginx/README.md.j2
Comment (alexandrevicenzi) [reply src/bci_build/package/nginx/README.md.j2:85]: Yes, same thing, just another path, but now I wonder, our default `nginx.conf` does not even load this stream path, just `vhosts.d/*.conf` and `conf.d/*.conf`.

Maybe not worth documenting this, unless we explicitly mention that this will only work if you change `nginx.conf`.

File: src/bci_build/package/nginx/README.md.j2
Comment (alexandrevicenzi) [reply src/bci_build/package/nginx/README.md.j2:55]: The **Using templates** section?

I guess it could come before **Environment variables** section, it makes more sense to me than after.

File: src/bci_build/package/nginx/README.md.j2
Comment (dmpop) [inline src/bci_build/package/nginx/README.md.j2:43]: Here's my suggestion.:

For example, if you want Nginx to use port 80, create a file named `/etc/nginx/templates/default.conf.template` containing the following variable definition:

```
listen       ${NGINX_PORT};
```

This makes the script to render the template above to `/etc/nginx/conf.d/default.conf` as follows:

```
listen       80;
```

File: src/bci_build/package/nginx/README.md.j2
Comment (alexandrevicenzi) [reply src/bci_build/package/nginx/README.md.j2:79]: The filter is used here:

```bash
awk "END { for (name in ENVIRON) { print ( name ~ /${filter}/ ) ? name : \"\" } }"
```

Honestly, I'm not sure what is allowed as I know little about `awk`.

File: src/bci_build/package/nginx/README.md.j2
Comment (dmpop) [inline src/bci_build/package/nginx/README.md.j2:59]:
```suggestion
This optional environment variable specifies a directory containing template files (default is `/etc/nginx/templates`).
```

File: src/bci_build/package/nginx/README.md.j2
Comment (dmpop) [inline src/bci_build/package/nginx/README.md.j2:61]:
```suggestion
**Note:** The script ignores template processing if this directory doesn't exist
```

File: src/bci_build/package/nginx/README.md.j2
Comment (dmpop) [inline src/bci_build/package/nginx/README.md.j2:65]:
```suggestion
This optional environment variable specifies a directory for storing results of running `envsubst` on templates (default is `/etc/nginx/conf.d`).
```

File: src/bci_build/package/nginx/README.md.j2
Comment (dmpop) [inline src/bci_build/package/nginx/README.md.j2:69]:
```suggestion
**Note:** Modifying this variable also requires changing the `nginx.conf`, so it recognizes the new directory location.
```

File: src/bci_build/package/nginx/README.md.j2
Comment (dmpop) [inline src/bci_build/package/nginx/README.md.j2:73]:
```suggestion
This optional environment variable changes the suffix of template files (default is `.template`).
```

File: src/bci_build/package/nginx/README.md.j2
Comment (dmpop) [inline src/bci_build/package/nginx/README.md.j2:75]:
```suggestion
**Note:** The script only processes files that have the specified suffix in their names.
```

File: src/bci_build/package/nginx/README.md.j2
Comment (dmpop) [inline src/bci_build/package/nginx/README.md.j2:79]:
```suggestion
This optional environment variable enables filtering out the processing of files in the location defined in `NGINX_ENVSUBST_TEMPLATE_DIR`.
```

File: src/bci_build/package/nginx/README.md.j2
Comment (dmpop) [inline src/bci_build/package/nginx/README.md.j2:83]:
```suggestion
To use scripts to perform configuration actions, add one or more `*.envsh` or `*.sh` scripts under `/docker-entrypoint.d/`. Any executable `*.envsh` or `*.sh` script found in the directory is executed before starting the service,  which can be used to perform further configuration steps.
```

File: src/bci_build/package/nginx/README.md.j2
Comment (dmpop) [inline src/bci_build/package/nginx/README.md.j2:88]:
```suggestion
- `20-envsubst-on-templates.sh` - Enables the use of environment variables in templates.
- `30-tune-worker-processes.sh` - Enables autotuning the number of worker processes.
```

File: src/bci_build/package/nginx/README.md.j2
Comment (dcermak) [reply src/bci_build/package/nginx/README.md.j2:79]: In the end that's just a regex match of `filter` (i.e. `NGINX_ENVSUBST_FILTER` or nothing) against the environment variable names. E.g. in my session:
```ShellSession
$ filter="SESSION"
$ printf '${%s} ' $(awk "END { for (name in ENVIRON) { print ( name ~ /${filter}/ ) ? name : \"\" } }" < /dev/null)
${DBUS_SESSION_BUS_ADDRESS} ${XDG_SESSION_PATH} ${GDMSESSION} ${XDG_SESSION_ID} ${DESKTOP_SESSION} ${XDG_SESSION_CLASS} ${STARSHIP_SESSION_KEY} ${XDG_SESSION_TYPE} ${XDG_SESSION_DESKTOP}

$ filter="[0-9]+"
$ printf '${%s} ' $(awk "END { for (name in ENVIRON) { print ( name ~ /${filter}/ ) ? name : \"\" } }" < /dev/null)
${GUESTFISH_PS1} ${I3SOCK} ${GTK3_MODULES}

Comment (alexandrevicenzi) [reply src/bci_build/package/nginx/README.md.j2:79]: Yes, indeed.

The current description is wrong, this variable allows you to filter out ENV variables.
If you set a value to it, any ENV variables that don't match will be omited and it won't be replaced in the template.

File: src/bci_build/package/nginx/README.md.j2
Comment (alexandrevicenzi) [reply src/bci_build/package/nginx/README.md.j2:45]: Ok.

File: src/bci_build/package/nginx/README.md.j2
Comment (alexandrevicenzi) [reply src/bci_build/package/nginx/README.md.j2:43]: Yes, better indeed.

File: src/bci_build/package/appcontainers.py
Comment (alexandrevicenzi) [reply src/bci_build/package/appcontainers.py:461]: Yes, the README version is never replaced, but removing this won't fix the `%%nginx_version%%` in the README by itself.

Should I hardcode versions in the code? Nginx version on TW changes frequently.

File: src/bci_build/package/nginx/README.md.j2
Comment (alexandrevicenzi) [reply src/bci_build/package/nginx/README.md.j2]: I updated this sentence.

File: src/bci_build/package/nginx/README.md.j2
Comment (alexandrevicenzi) [reply src/bci_build/package/nginx/README.md.j2]: I removed this part, not worth having as our default config does not support it.

File: src/bci_build/package/nginx/README.md.j2
Comment (alexandrevicenzi) [reply src/bci_build/package/nginx/README.md.j2:7]: I can.

File: src/bci_build/package/appcontainers.py
Comment (dcermak) [reply src/bci_build/package/appcontainers.py]: It has been merged now, please rebase and adjust to use the hardcoded version

File: src/bci_build/package/nginx/README.md.j2
Comment (alexandrevicenzi) [reply src/bci_build/package/nginx/README.md.j2:7]: The issue is that some places use Nginx (e.g Wikipedia), nginx (official nginx.org docs), and NGINX (F5 nginx.com website).

The DockerHub image mixes all 3 naming versions. Not sure if there's an agreement on which one is the official, but all lower case seems to be more popular.

File: src/bci_build/package/appcontainers.py
Comment (alexandrevicenzi) [reply src/bci_build/package/appcontainers.py]: @dcermak I fixed this using the new `package_versions.json` file.

File: src/bci_build/package/nginx/README.md.j2
Comment (dcermak) [inline src/bci_build/package/nginx/README.md.j2:53]: What is the default?

File: src/bci_build/package/nginx/README.md.j2
Comment (dcermak) [inline src/bci_build/package/nginx/README.md.j2:31]:
```suggestion
The script reads `.template` files stored in `/etc/nginx/templates/` and saves the result of the [`envsubst`](https://www.gnu.org/software/gettext/manual/html_node/envsubst-Invocation.html) command to the directory `/etc/nginx/conf.d/`.
```

**Review by dcermak — CHANGES_REQUESTED** [review]
LGTM! Just some nitpicks and please also add a `generate_package_version_check("nginx", nginx_version)` into the `custom_end` string.

File: src/bci_build/package/nginx/README.md.j2
Comment (alexandrevicenzi) [reply src/bci_build/package/nginx/README.md.j2]: Disabled, I added this info.

File: src/bci_build/package/nginx/README.md.j2
Comment (alexandrevicenzi) [reply src/bci_build/package/nginx/README.md.j2]: Fixed.

File: src/bci_build/package/nginx/README.md.j2
Comment (alexandrevicenzi) [reply src/bci_build/package/nginx/README.md.j2]: Fixed.

File: src/bci_build/package/nginx/README.md.j2
Comment (alexandrevicenzi) [reply src/bci_build/package/nginx/README.md.j2]: Everything is lower case now.

File: src/bci_build/package/nginx/README.md.j2
Comment (dcermak) [inline src/bci_build/package/nginx/README.md.j2:39]: I don't understand this sentence, could you reword it please?

Comment (alexandrevicenzi) [reply src/bci_build/package/nginx/README.md.j2:39]: > The template above is then rendered to `/etc/nginx/conf.d/default.conf` as follows:

Is this version understandable?

Comment (alexandrevicenzi) [reply src/bci_build/package/nginx/README.md.j2]: Fixed.

**Review by dcermak — DISMISSED** [review]
Thanks a lot for your patience and addressing all the feedback!

**Comment by dcermak** [comment]
Fixed missing import of `generate_package_version_check`

**Pull request merged**
